src/graph/graphcontext.c

```diff
@@ -111,84 +87,164 @@ LabelStore* GraphContext_GetStore(const GraphContext *gc, const char *label, Lab
+  // Property does not exist or was not indexed.
+  if (idx == NULL) return INDEX_FAIL;
+
+  // Remove the index association from the label store
+  LabelStore_AssignValue(store, (char*)property, NULL);
+  // Index *idx = gc->indices[offset];
+
+  // Pop the last stored index
+  Index *last_idx = array_pop(gc->indices);
+  if (idx != last_idx) {
+    // If the index being deleted is not the last, swap the last into the newly-emptied position
+    uint32_t pos;
+    // Find the position of the index in the gc->indices array
+    for (pos = 0; pos < array_len(gc->indices); pos ++) {
+      if (gc->indices[pos] == idx) {
+        gc->indices[pos] = last_idx;
+        break;
+      }
+    }
+    // Fail if the index was somehow not in the indices array.
+    assert(pos < array_len(gc->indices));
+  }
 
-  Index *idx = gc->indices[offset];
+  // Free the index
   Index_Free(idx);
-  gc->index_count --;
-  // If the deleted index was not the last, swap the last into the newly-emptied position
-  if (offset < gc->index_count) gc->indices[offset] = gc->indices[gc->index_count];
 
   return INDEX_OK;
 }
 
+// Add references to a node to all indices built upon its properties
+void GraphContext_AddNodeToIndices(GraphContext *gc, LabelStore *store, Node *n) {
+  if (store && GraphContext_HasIndices(gc)) {
+    EntityProperty *props = ENTITY_PROPS(n);
+    for (int j = 0; j < ENTITY_PROP_COUNT(n); j ++) {
+      // If a property is indexed, update it to include the given node 
+      Index *idx = _GraphContext_GetIndexFromStore(store, props[j].name);
+      if (idx) Index_InsertNode(idx, n->entity->id, &props[j].value);
+    }
+  }
+}
+
+// Delete all references to a node from any indices built upon its properties
+void GraphContext_DeleteNodeFromIndices(GraphContext *gc, LabelStore *store, Node *n) {
+  if (!GraphContext_HasIndices(gc)) return;
+
+  if (store == NULL) {
+    if (n->label) {
+      // Node will have a label string if one was specified in the query MATCH clause
+      store = GraphContext_GetStore(gc, n->label, STORE_NODE);
+    } else {
+      // Otherwise, look up the offset of the matching label (if any)
+      int store_id = Graph_GetLabelID(gc->g, n->entity->id);
+      // Do nothing if node had no label
+      if (store_id == GRAPH_NO_LABEL) return;
+      store = gc->node_stores[store_id];
+    }
+  }
+
+  // Update any indices this entity is represented in
+  EntityProperty *props = ENTITY_PROPS(n);
+  for (int j = 0; j < ENTITY_PROP_COUNT(n); j ++) {
+    Index *idx = LabelStore_RetrieveValue(store, props[j].name);
+    if (idx != TRIEMAP_NOTFOUND && idx != NULL) {
+      Index_DeleteNode(idx, n->entity->id, &props[j].value);
+    }
+  }
+}
+
+// If the given property is indexed, update node reference within index to point to correct value
+void GraphContext_UpdateNodeIndices(GraphContext *gc, LabelStore *store, NodeID id, EntityProperty *prop, SIValue *newval) {
```

Comment:
Would it be possible to pass in the Node object, such that in cases we have its label we won't have to call Graph_GetLabelID which is a costly operation

Reply:
The SET operation doesn't ever retrieve Node or Edge objects; it works exclusively with datablocks. An optimization I've added is that if the MATCH clause specifies a label, `store` will always be provided, but if we have no label and just the ID, I see no alternative to this approach.